File: leetcode/15.go
//三数之和
import "sort"

func threeSum(nums []int) [][]int {
	res := [][]int{}
	sort.Ints(nums)

	for i := 0; i < len(nums) - 2; i++ {
   	n1 := nums[i]
		if n1 > 0 {
			break
		}

		if i > 0 && nums[i] == nums[i-1] {
			continue
		}	
		
		l, r := i+1, len(nums)-1
		for l < r {
			n2, n3 := nums[l], nums[r]
			if n1+n2+n3 == 0 {
				res = append(res, []int{n1, n2, n3})
				for l<r && n2==nums[l] {
					l++
				}
				for l<r && n3==nums[r] {
					r--
				}
			} else if n1+n2+n3 >0 {
				r--
			} else {
				l++
			}
		}
  }
	return res
}

File: array/leetcode/cpp/15.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> result;
        sort(nums.begin(), nums.end());

        for (int i = 0; i < nums.size(); i++) {
            if (i > 0 && nums[i] == nums[i-1]) {
                continue;
            }
            int left = i+1;
            int right = nums.size()-1;
            while(right>left) {
                if (nums[i] + nums[left] + nums[right] == 0) {
                    result.push_back(vector<int>{nums[i], nums[left], nums[right]});
                    while (right > left && nums[right] == nums[right-1]) right--;
                    while (right > left && nums[left] == nums[left+1]) left++;
                    right--;
                    left++;
                } else if (nums[i] + nums[left] + nums[right] > 0) {
                    right--;
                } else {
                    left++;
                }
            }

        }

        return result;
    }
};